tests/test_project.py

```diff
@@ -1775,6 +1776,178 @@ class TestGroupDynamicProjectMainInterface(TestProjectMainInterface):
+        assert agg_cursor._project == project
+        assert agg_cursor._filter is None
+        assert agg_cursor._doc_filter is None
+        # Since default aggregate is not present in project, hence no aggregator
+        # will be evaulated as equal and hence all the aggregates present in this project
+        # will be considered unique.
+        assert len(agg_cursor) == 41
+        assert tuple(project) in agg_cursor
+
+    def test_filters(self):
+        project = self.mock_project()
+        agg_cursor = _AggregatesCursor(project=project, filter={"even": True})
+        assert agg_cursor._project == project
+        assert agg_cursor._filter == {"even": True}
+        assert agg_cursor._doc_filter is None
+        assert len(agg_cursor) == 15
+
+    def test_reregister_aggregates(self):
+        project = self.mock_project()
+        agg_cursor = _AggregatesCursor(project=project)
+        assert len(agg_cursor) == 41
+        project.open_job(dict(i=31, even=False)).init()
+        # Default aggregate store doesn't need to be re-registered.
+        assert len(agg_cursor) == 42
+        project.reregister_aggregates()
+        # The operation agg_op2 adds another aggregate in the project.
+        assert len(agg_cursor) == 43
+
+
+class TestAggregationProjectMainInterface(TestAggregatesProjectBase):
+    def test_main_run(self):
+        project = self.mock_project()
+        assert len(project)
+        for job in project:
+            assert not job.doc.get("sum", False)
+            assert not job.doc.get("sum_other", False)
+            assert not job.doc.get("sum_custom", False)
+            assert not job.doc.get("op2", False)
+            assert not job.doc.get("op3", False)
+
+        even_sum = sum([job.sp.i for job in project if job.sp.i % 2 == 0])
+        odd_sum = sum([job.sp.i for job in project if job.sp.i % 2 != 0])
+
+        self.call_subcmd(
+            "run -o agg_op1 agg_op1_different agg_op1_custom agg_op2 agg_op3 --show-traceback"
+        )
+
+        for job in project:
+            assert job.doc.op2
+            assert job.doc.op3
+            if job.sp.i % 2 == 0:
+                assert (
+                    job.doc.sum == job.doc.sum_other == job.doc.sum_custom == even_sum
+                )
+            else:
+                assert job.doc.sum == job.doc.sum_other == job.doc.sum_custom == odd_sum
+
+    def test_invalid_with_job(self):
+        project = self.mock_project()
+        assert len(project)
+
+        with pytest.raises(Exception) as runinfo:
+            project.run(names=["agg_op5"])
+            assert (
+                "use of @with_job decorator with aggregation" in runinfo.value.message
+            )
+
+    def test_main_run_cmd(self):
+        project = self.mock_project()
+        assert len(project)
+
+        run_output = self.call_subcmd("run -o agg_op4").decode().splitlines()
+
+        assert "1 and 2" in "\n".join(run_output)
```

Comment:
I think you can do this instead of splitting/joining. I recently made this change in some other tests. Match this change (remove `splittlines` and `join`) in the other tests below, please.
```suggestion
        run_output = self.call_subcmd("run -o agg_op4").decode("utf-8")

        assert "1 and 2" in run_output
```